commands/dht.go

```diff
@@ -0,0 +1,163 @@
+
+// Note, most of this is copied directly from go-ipfs (https://github.com/ipfs/go-ipfs/blob/master/core/commands/dht.go).
+// A few simple modifications have been adapted for filecoin.
+var dhtCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline:          "Issue commands directly through the DHT.",
+		ShortDescription: ``,
+	},
+
+	Subcommands: map[string]*cmds.Command{
+		"findprovs": findProvidersDhtCmd,
+	},
+}
+
+var findProvidersDhtCmd = &cmds.Command{
+	Helptext: cmdkit.HelpText{
+		Tagline:          "Find peers that can provide a specific value, given a key.",
+		ShortDescription: "Outputs a list of newline-delimited provider Peer IDs.",
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("key", true, false, "The key to send provide records for.").EnableStdin(),
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption(dhtVerboseOptionName, "v", "Print extra information."),
+		cmdkit.IntOption(numProvidersOptionName, "n", "The number of providers to find.").WithDefault(20),
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+		numProviders, _ := req.Options[numProvidersOptionName].(int)
+		if numProviders < 1 {
+			return fmt.Errorf("number of providers must be greater than 0")
+		}
+
+		c, err := cid.Parse(req.Arguments[0])
+		if err != nil {
+			return err
+		}
+
+		ctx, cancel := context.WithCancel(req.Context)
+		ctx, events := notif.RegisterForQueryEvents(ctx)
+
+		pchan := GetPorcelainAPI(env).RouterFindProvidersAsync(ctx, c, numProviders)
+
+		go func() {
+			defer cancel()
+			for p := range pchan {
+				np := p
+				notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+					Type:      notif.Provider,
+					Responses: []*pstore.PeerInfo{&np},
+				})
+			}
+		}()
+		for e := range events {
+			if err := res.Emit(e); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeTypedEncoder(func(req *cmds.Request, w io.Writer, out *notif.QueryEvent) error {
+			pfm := pfuncMap{
+				notif.FinalPeer: func(obj *notif.QueryEvent, out io.Writer, verbose bool) {
+					if verbose {
+						fmt.Fprintf(out, "* closest peer %s\n", obj.ID) // nolint: errcheck
+					}
+				},
+				notif.Provider: func(obj *notif.QueryEvent, out io.Writer, verbose bool) {
+					prov := obj.Responses[0]
+					if verbose {
+						fmt.Fprintf(out, "provider: ") // nolint: errcheck
+					}
+					fmt.Fprintf(out, "%s\n", prov.ID.Pretty()) // nolint: errcheck
+					if verbose {
+						for _, a := range prov.Addrs {
+							fmt.Fprintf(out, "\t%s\n", a) // nolint: errcheck
+						}
+					}
+				},
+			}
+
+			verbose, _ := req.Options[dhtVerboseOptionName].(bool)
+			printEvent(out, w, verbose, pfm)
+
+			return nil
+		}),
+	},
+	Type: notif.QueryEvent{},
+}
+
+type printFunc func(obj *notif.QueryEvent, out io.Writer, verbose bool)
+type pfuncMap map[notif.QueryEventType]printFunc
+
+func printEvent(obj *notif.QueryEvent, out io.Writer, verbose bool, override pfuncMap) {
```

Comment:
Note that this function is only needed to enable the output logging of events that only show up during a "verbose" run.  If we choose to eliminate the verbose option the code will get significantly shorter.  However if we keep the verbose option on the other (to be written) dht subcommands we can reuse this function and the overhead goes way down.

Reply:
Move this comment into the code.

Reply:
good idea